cmd/gardener-scheduler/app/gardener_scheduler.go

```diff
@@ -0,0 +1,373 @@
+}
+
+// NewGardenerScheduler is the main entry point of instantiating a new Gardener Scheduler.
+func NewGardenerScheduler(cfg *config.SchedulerConfiguration) (*GardenerScheduler, error) {
+	if cfg == nil {
+		return nil, errors.New("config is required")
+	}
+
+	// validate the configuration
+	if err := validation.ValidateConfiguration(cfg); err != nil {
+		return nil, err
+	}
+
+	// Initialize logger
+	logger := logger.NewLogger(cfg.LogLevel)
+	logger.Info("Starting Gardener scheduler ...")
+
+	if flag := flag.Lookup("v"); flag != nil {
+		if err := flag.Value.Set(fmt.Sprintf("%d", cfg.KubernetesLogLevel)); err != nil {
+			return nil, err
+		}
+	}
+
+	// Prepare a Kubernetes client object for the Garden cluster which contains all the Clientsets
+	// that can be used to access the Kubernetes API.
+	if kubeconfig := os.Getenv("KUBECONFIG"); kubeconfig != "" {
+		cfg.ClientConnection.Kubeconfig = kubeconfig
+	}
+
+	restCfg, err := server.RestConfigFromClientConnectionConfiguration(cfg.ClientConnection)
+	if err != nil {
+		return nil, err
+	}
+
+	disc, err := discoveryFromSchedulerConfiguration(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	k8sGardenClient, err := kubernetes.NewForConfig(restCfg, client.Options{
+		Mapper: restmapper.NewDeferredDiscoveryRESTMapper(disc),
+		Scheme: kubernetes.GardenScheme,
+	})
+	if err != nil {
+		return nil, err
+	}
+	k8sGardenClientLeaderElection, err := k8s.NewForConfig(restCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up leader election if enabled and prepare event recorder.
+	var (
+		leaderElectionConfig *leaderelection.LeaderElectionConfig
+		recorder             = server.CreateRecorder(k8sGardenClient.Kubernetes(), "gardener-scheduler")
+	)
+	if cfg.LeaderElection.LeaderElect {
+		leaderElectionConfig, err = server.MakeLeaderElectionConfig(cfg.LeaderElection.LeaderElectionConfiguration, cfg.LeaderElection.LockObjectNamespace, cfg.LeaderElection.LockObjectName, k8sGardenClientLeaderElection, recorder)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &GardenerScheduler{
+		Config:                 cfg,
+		Logger:                 logger,
+		Recorder:               recorder,
+		K8sGardenClient:        k8sGardenClient,
+		K8sGardenInformers:     gardeninformers.NewSharedInformerFactory(k8sGardenClient.Garden(), 0),
+		K8sGardenCoreInformers: gardencoreinformers.NewSharedInformerFactory(k8sGardenClient.GardenCore(), 0),
+		KubeInformerFactory:    kubeinformers.NewSharedInformerFactory(k8sGardenClient.Kubernetes(), 0),
+		LeaderElection:         leaderElectionConfig,
+	}, nil
+}
+
+func (g *GardenerScheduler) cleanup() {
+	if err := os.RemoveAll(schedulerconfigv1alpha1.DefaultDiscoveryDir); err != nil {
+		g.Logger.Errorf("Could not cleanup base discovery cache directory: %v", err)
+	}
+}
+
+// Run runs the Gardener Scheduler. This should never exit.
+func (g *GardenerScheduler) Run(ctx context.Context, cancel context.CancelFunc) error {
```

Comment:
Why does this thing get a `context.CancelFunc`? It should only get a `context.Context` and the 'supervising entity' should cancel all others as soon as this errors.

Reply:
What do you mean by "should cancel all others as soon as this errors"?

Isn't the cancle function being used to cancle the "main" context? Therefore we call the cancle() once we lose leadership - to stop the scheduler effectively.  
It is basically the same functionality [as in the controller manager](https://github.com/gardener/gardener/blob/c4f251b154b94de2443bc068c28aea8f6e4c681d/cmd/gardener-controller-manager/app/gardener_controller_manager.go#L363).
Why do we need a different solution for the scheduler? I think i do not quite understand the point you made above.

Reply:
Yes I see but maybe then already back then we've done a design mistake. A rule of thumb is that an entity should only `cancel` another entity if it was the one that created it in order to have a 'supervision-graph-like' thing. In this case, the function is just calling a `cancel` function which it doesn't know anything about. A better design would be if this function sets up its own `cancel` at the beginning using its parent `ctx` and `context.WithCancel` and then once leadership is lost it calls `cancel` on that function. You're almost doing this with the `leaderElectionCtx, leaderElectionCancel` however you're not using the obtained contexts for the functions to run with. Just use this and you should be fine. If you have any questions about how to fix this, just reach out to me and we can clarify this outside this GitHub issue.